src/main/java/com/aws/greengrass/mqttclient/AwsIotMqttClient.java

```diff
@@ -222,26 +250,27 @@ synchronized boolean connected() {
     }
 
     @SuppressWarnings("PMD.NullAssignment")
-    private synchronized void disconnect() {
-        try {
-            currentlyConnected.set(false);
-            if (connection != null) {
-                logger.atDebug().log("Disconnecting from AWS IoT Core");
-                try {
-                    connection.disconnect().get(getTimeout(), TimeUnit.MILLISECONDS);
-                } finally {
+    protected synchronized CompletableFuture<Void> disconnect() {
+        currentlyConnected.set(false);
+        if (connection != null) {
+            logger.atDebug().log("Disconnecting from AWS IoT Core");
+            return connection.disconnect().whenComplete((future, error) -> {
+                logger.atDebug().log("Successfully disconnected from AWS IoT Core");
+                synchronized (this) {
                     connection.close();
                     connection = null;
```

Comment:
might want to reset the connection future here too.

Reply:
we restrict the connect() operation such that there is only one ongoing connect at any particular time. If connect is ongoing and disconnect is called, resetting the connectionFuture here will allow another connect operation to go through when the previous one is still in progress. Ideally we should not do that unless I am missing something

Reply:
Also connect and disconnect are protected so we have control on how they are called.